### PR TITLE
Implement Pub/Sub API using SSE via NATS

### PR DESCRIPTION
**Pull request by Linoy339**
- Listener api

- Publishing changes to apis

### CODE REVIEW

**Comment by avaidyam** [comment]
@Linoy339 @michaelmenon 
For the client to begin listening, it looks like there's one specific `/listen` endpoint -- would it be possible to change this? The API requires it to be `SUBSCRIBE <topic_name>` so perhaps in express we should do `app.subscribe('*', ...)` instead. Could we make it so instead of e.g. "GET /listen/study/participant/?study_id=345&participant_id=123" it would be e.g. "SUBSCRIBE /study/345/participant/123/"?

**Comment by Linoy339** [comment]
@avaidyam. Its changed from listen to SUBSCRIBE.

**Comment by avaidyam** [comment]
@Linoy339 @michaelmenon This isn't the change I mentioned... the HTTP Method is `SUBSCRIBE`, and the HTTP Path should be `*`. Please fix that, thanks.

**Comment by Linoy339** [comment]
@avaidyam . We have pushed the changes which included the subscribe http method instead of get along with push notification api for the external clients to call.

**Comment by avaidyam** [comment]
@Linoy339 What are the related environment variables that need to be set correctly? And is there any specific NATS configuration required? I'll be bringing this up for testing in staging/production this week.

**Comment by Linoy339** [comment]
Nats server should be installed: [https://nats.io/download/nats-io/nats-server/](https://nats.io/download/nats-io/nats-server/)

The env variable is NATSServers. This variable is kept to carry multiple nats server address(if present). We can do it by putting all server address as comma seperated.
For now, we assume that only one nats server is present. (eg: NATSServers=0.0.0.0:4222)

**Comment by avaidyam** [comment]
@Linoy339 Can you change it back to `GET /subscribe` based on what @michaelmenon [mentioned in this issue](https://github.com/BIDMCDigitalPsychiatry/LAMP-platform/issues/6)? And can you rename the environment variable to be exactly: `NATS_SERVER`? Otherwise, looks good!

**Comment by Linoy339** [comment]
@avaidyam. Its done.

File: src/utils/PushNotificationAPI.ts [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+import { SensorEventRepository, ParticipantRepository, TypeRepository } from "../repository"
+import { PushNotificationQueue } from "../utils/queue/PushNotificationQueue"
+
+export const PushNotificationAPI = Router()
+
+/**Push notification API--used by external clients to send notifications
+ *
+ *
+ */
+PushNotificationAPI.post("/notifications", async (req: Request, res: Response) => {
```
Comment (avaidyam) [inline src/utils/PushNotificationAPI.ts:12]: Can you add some documentation around what the input and output are for this API call?

**Review by avaidyam — APPROVED** [review]
Looks good but a few remaining questions before we merge and push!

File: src/service/ActivityEventService.ts [hunk 2]
```diff
@@ -20,6 +21,35 @@ ActivityEventService.post("/participant/:participant_id/activity_event", async (
...
+      topic: `participant.*.activity_event`,
+      action: "create",
+      participant_id: participant_id,
+      payload: Array.isArray(activity_event) ? activity_event : [activity_event],
+    })
+
+     PubSubAPIListenerQueue.add({
+      topic: `activity.*.activity_event`,
+      action: "create",
+      participant_id: participant_id,
```
[the remark this answers is not in the document]
Comment (Linoy339) [reply src/service/ActivityEventService.ts:43]: activity_id will be present in the activity_event array which is given as payload here. Its  dispatched to PubSubAPIListenerQueue. From there, the payload gets processed. i.e activity_event array is looped and published

File: src/service/ActivityEventService.ts [hunk 3]
```diff
@@ -20,6 +21,35 @@ ActivityEventService.post("/participant/:participant_id/activity_event", async (
...
+      payload: Array.isArray(activity_event) ? activity_event : [activity_event],
+    })
+
+    PubSubAPIListenerQueue.add({
+      topic: `participant.*.activity.*.activity_event`,
+      action: "create",
+      participant_id: participant_id,
+      payload: Array.isArray(activity_event) ? activity_event : [activity_event],
```
[the remark this answers is not in the document]
Comment (Linoy339) [reply src/service/ActivityEventService.ts:51]: Yes. Its done individually, but from  PubSubAPIListenerQueue consumer
Here, the whole activity_event array is passed to PubSubAPIListenerQueue as payload and treated individually by looping it from the queue consumer. see [https://github.com/BIDMCDigitalPsychiatry/LAMP-server/blob/nats-publish/src/utils/queue/PubSubAPIListenerQueue.ts#L67](https://github.com/BIDMCDigitalPsychiatry/LAMP-server/blob/nats-publish/src/utils/queue/PubSubAPIListenerQueue.ts#L63)

**Comment by avaidyam** [comment]
Thanks for the responses @Linoy339 - they are very helpful. Going forwards could you add lots of comments to the code itself to clarify these things ahead of time? I think we're ready to merge this PR though!

File: src/service/ActivityService.ts [hunk 4]
```diff
@@ -14,6 +15,13 @@ ActivityService.post("/study/:study_id/activity", async (req: Request, res: Resp
     const activity = req.body
     study_id = await _verify(req.get("Authorization"), ["self", "sibling", "parent"], study_id)
     const output = { data: await ActivityRepository._insert(study_id, activity) }
+    activity.study_id = study_id
+    activity.action = "create"
+
+    //publishing data
+    PubSubAPIListenerQueue.add({ topic: `activity`, token: `study.${study_id}.activity.${output['data']}`, payload: activity })
```
[the remark this answers is not in the document]
Comment (Linoy339) [reply src/service/ActivityService.ts:22]: In ActivityService.ts, we can create token easily, as its a request of single object. So we can form the token from ActivityService.ts itself.
study_id will be given in the input request and the activity_id created is gettable  from " output['data'] "

For activity_event,  we will convert into array if the request is not array formatted. Its difficult to create token  from ActivityEventService.ts, as  its changed  into array.  So, we pass the whole array and create token from PubSubAPIListenerQueue Consumer .
Also, participant.*.activity_event is the topic which we subscribe internally. Based on this, we create token and sent to the consumer.
See [https://github.com/BIDMCDigitalPsychiatry/LAMP-server/blob/nats-publish/src/utils/queue/PubSubAPIListenerQueue.ts#L74](https://github.com/BIDMCDigitalPsychiatry/LAMP-server/blob/nats-publish/src/utils/queue/PubSubAPIListenerQueue.ts#L74)

**Comment by Linoy339** [comment]
@avaidyam .  I can detail it more and push the code.

**Comment by avaidyam** [comment]
@Linoy339 Please do! As much as you can add in comments to the push notifications code as well as this code will be extremely helpful. Thanks!

**Comment by Linoy339** [comment]
@avaidyam .  Please check [https://github.com/BIDMCDigitalPsychiatry/LAMP-server/pull/34](https://github.com/BIDMCDigitalPsychiatry/LAMP-server/pull/34)

**Pull request merged**